File: parsec/core/trustchain.py
# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPL-3.0 2016-present Scille SAS

from typing import TYPE_CHECKING, Tuple, List, Sequence, Optional
from parsec._parsec import DateTime

from parsec.crypto import VerifyKey
from parsec.api.protocol import UserID, DeviceID
from parsec.api.data import (
    DataError,
    UserCertificateContent,
    UserProfile,
    RevokedUserCertificateContent,
    DeviceCertificateContent,
)


class TrustchainError(Exception):
    pass


def _build_signature_path(*devices_ids):
    return " <-sign- ".join([str(d) for d in devices_ids])


class CertifState:
    __slots__ = ("certif", "content", "verified")

    def __init__(self, certif, content, verified):
        self.certif = certif
        self.content = content
        self.verified = verified


class TrustchainContext:
    def __init__(self, root_verify_key: VerifyKey, cache_validity: int):
        self.root_verify_key = root_verify_key
        self.cache_validity = cache_validity
        self._users_cache = {}
        self._devices_cache = {}
        self._revoked_users_cache = {}

    def invalidate_user_cache(self, user_id: UserID) -> None:
        self._users_cache.pop(user_id, None)

    def get_user(self, user_id: UserID, now: DateTime = None) -> Optional[UserCertificateContent]:
        now = now or DateTime.now()
        try:
            cached_on, verified_user = self._users_cache[user_id]
            if now - cached_on < self.cache_validity:
                return verified_user
        except KeyError:
            pass
        return None

    def get_revoked_user(
        self, user_id: UserID, now: DateTime = None
    ) -> Optional[RevokedUserCertificateContent]:
        now = now or DateTime.now()
        try:
            cached_on, verified_revoked_user = self._revoked_users_cache[user_id]
            if now - cached_on < self.cache_validity:
                return verified_revoked_user
        except KeyError:
            pass
        return None

    def get_device(
        self, device_id: DeviceID, now: DateTime = None
    ) -> Optional[DeviceCertificateContent]:
        now = now or DateTime.now()
        try:
            cached_on, verified_device = self._devices_cache[device_id]
            if now - cached_on < self.cache_validity:
                return verified_device
        except KeyError:
            pass
        return None

    def load_user_and_devices(
        self,
        trustchain: dict,
        user_certif: bytes,
        revoked_user_certif: Optional[bytes] = None,
        devices_certifs: Sequence[bytes] = (),
        expected_user_id: UserID = None,
    ) -> Tuple[
        UserCertificateContent,
        Optional[RevokedUserCertificateContent],
        List[DeviceCertificateContent],
    ]:
        now = DateTime.now()
        verified_users, verified_revoked_users, verified_devices = self.load_trustchain(
            users=(user_certif, *trustchain["users"]),
            revoked_users=(revoked_user_certif, *trustchain["revoked_users"])
            if revoked_user_certif
            else trustchain["revoked_users"],
            devices=(*devices_certifs, *trustchain["devices"]),
            now=now,
        )

        verified_user = verified_users[0]
        if expected_user_id and verified_user.user_id != expected_user_id:
            raise TrustchainError(
                f"Unexpected certificate: expected `{expected_user_id}` but got `{verified_user.user_id}`"
            )

        if revoked_user_certif:
            verified_revoked_user = verified_revoked_users[0]
            if expected_user_id and verified_revoked_user.user_id != expected_user_id:
                raise TrustchainError(
                    f"Unexpected certificate: expected `{expected_user_id}` but got `{verified_revoked_user.user_id}`"
                )
        else:
            verified_revoked_user = None

        verified_devices = verified_devices[: len(devices_certifs)]
        for verified_device in verified_devices:
            if expected_user_id and verified_device.device_id.user_id != expected_user_id:
                raise TrustchainError(
                    f"Unexpected certificate: expected `{expected_user_id}` but got `{verified_device.device_id}`"
                )

        return verified_user, verified_revoked_user, verified_devices

    def load_trustchain(
        self,
        users: Sequence[bytes] = (),
        revoked_users: Sequence[bytes] = (),
        devices: Sequence[bytes] = (),
        now: DateTime = None,
    ) -> Tuple[
        List[UserCertificateContent],
        List[RevokedUserCertificateContent],
        List[DeviceCertificateContent],
    ]:
        now = now or DateTime.now()

        users_states = {}
        devices_states = {}
        revoked_users_states = {}

        # Deserialize the certificates and filter the ones we already have in cache
        try:
            for certif in devices:
                unverified_device = DeviceCertificateContent.unsecure_load(certif)
                verified_device = self.get_device(unverified_device.device_id, now)
                if verified_device:
                    devices_states[verified_device.device_id] = CertifState(
                        certif, verified_device, True
                    )
                else:
                    devices_states[unverified_device.device_id] = CertifState(
                        certif, unverified_device, False
                    )

            for certif in users:
                unverified_user = UserCertificateContent.unsecure_load(certif)
                verified_user = self.get_user(unverified_user.user_id, now)
                if verified_user:
                    users_states[verified_user.user_id] = CertifState(certif, verified_user, True)
                else:
                    users_states[unverified_user.user_id] = CertifState(
                        certif, unverified_user, False
                    )

            for certif in revoked_users:
                unverified_revoked_user = RevokedUserCertificateContent.unsecure_load(certif)
                verified_revoked_user = self.get_revoked_user(unverified_revoked_user.user_id, now)
                if verified_revoked_user:
                    revoked_users_states[verified_revoked_user.user_id] = CertifState(
                        certif, verified_revoked_user, True
                    )
                else:
                    revoked_users_states[unverified_revoked_user.user_id] = CertifState(
                        certif, unverified_revoked_user, False
                    )

        except DataError as exc:
            raise TrustchainError(f"Invalid certificate: {exc}") from exc

        def _get_eventually_verified_user(user_id):
            try:
                return users_states[user_id].content
            except KeyError:
                return None

        def _get_eventually_verified_revoked_user(user_id):
            try:
                return revoked_users_states[user_id].content
            except KeyError:
                return None

        def _verify_created_by_root(certif, certif_cls, sign_chain):
            try:
                return certif_cls.verify_and_load(
                    certif, author_verify_key=self.root_verify_key, expected_author=None
                )

            except DataError as exc:
                path = _build_signature_path(*sign_chain, "<Root Key>")
                raise TrustchainError(f"{path}: Invalid certificate: {exc}") from exc

        def _verify_created_by_device(certif, certif_cls, author_id, sign_chain):
            author_device = _recursive_verify_device(author_id, sign_chain)
            try:
                verified = certif_cls.verify_and_load(
                    certif,
                    author_verify_key=author_device.verify_key,
                    expected_author=author_device.device_id,
                )

            except DataError as exc:
                path = _build_signature_path(*sign_chain, author_id)
                raise TrustchainError(f"{path}: Invalid certificate: {exc}") from exc

            # Author is either admin or signing one of it own devices
            verified_user_id = (
                verified.device_id.user_id
                if isinstance(verified, DeviceCertificateContent)
                else verified.user_id
            )
            if author_device.device_id.user_id != verified_user_id:
                author_user = _get_eventually_verified_user(author_id.user_id)
                if not author_user:
                    path = _build_signature_path(*sign_chain, author_id)
                    raise TrustchainError(
                        f"{path}: Missing user certificate for {author_id.user_id}"
                    )
                elif author_user.profile != UserProfile.ADMIN:
                    path = _build_signature_path(*sign_chain, author_id)
                    raise TrustchainError(
                        f"{path}: Invalid signature given {author_user.user_id} is not admin"
                    )
            # Also make sure author wasn't revoked at creation time
            author_revoked_user = _get_eventually_verified_revoked_user(author_id.user_id)
            if author_revoked_user and verified.timestamp > author_revoked_user.timestamp:
                path = _build_signature_path(*sign_chain, author_id)
                raise TrustchainError(
                    f"{path}: Signature ({verified.timestamp}) is posterior "
                    f"to user revocation ({author_revoked_user.timestamp})"
                )

            return verified

        def _recursive_verify_device(device_id, signed_children=()):
            if device_id in signed_children:
                path = _build_signature_path(*signed_children, device_id)
                raise TrustchainError(f"{path}: Invalid signature loop detected")

            try:
                state = devices_states[device_id]
            except KeyError:
                path = _build_signature_path(*signed_children, device_id)
                raise TrustchainError(f"{path}: Missing device certificate for {device_id}")

            author = state.content.author
            if author is None:
                verified_device = _verify_created_by_root(
                    state.certif, DeviceCertificateContent, sign_chain=(*signed_children, device_id)
                )
            else:
                verified_device = _verify_created_by_device(
                    state.certif,
                    DeviceCertificateContent,
                    author,
                    sign_chain=(*signed_children, device_id),
                )
            return verified_device

        def _verify_user(unverified_content, certif):
            author = unverified_content.author
            user_id = unverified_content.user_id
            if author is None:
                verified_user = _verify_created_by_root(
                    certif, UserCertificateContent, sign_chain=(f"{user_id}'s creation",)
                )
            elif author.user_id == user_id:
                raise TrustchainError(f"{user_id}: Invalid self-signed user certificate")
            else:
                verified_user = _verify_created_by_device(
                    certif, UserCertificateContent, author, sign_chain=(f"{user_id}'s creation",)
                )
            return verified_user

        def _verify_revoked_user(unverified_content, certif):
            author = unverified_content.author
            user_id = unverified_content.user_id
            if author is None:
                verified_revoked_user = _verify_created_by_root(
                    certif, RevokedUserCertificateContent, sign_chain=(f"{user_id}'s revocation",)
                )
            elif author.user_id == user_id:
                raise TrustchainError(f"{user_id}: Invalid self-signed user revocation certificate")
            else:
                verified_revoked_user = _verify_created_by_device(
                    certif,
                    RevokedUserCertificateContent,
                    author,
                    sign_chain=(f"{user_id}'s revocation",),
                )
            return verified_revoked_user

        # Verified what need to be and populate the cache with them

        for certif_state in devices_states.values():
            if not certif_state.verified:
                certif_state.content = _recursive_verify_device(certif_state.content.device_id)
        for certif_state in users_states.values():
            if not certif_state.verified:
                certif_state.content = _verify_user(certif_state.content, certif_state.certif)
        for certif_state in revoked_users_states.values():
            if not certif_state.verified:
                certif_state.content = _verify_revoked_user(
                    certif_state.content, certif_state.certif
                )

        # Finally populate the cache
        for certif_state in devices_states.values():
            if not certif_state.verified:
                self._devices_cache[certif_state.content.device_id] = (now, certif_state.content)
        for certif_state in users_states.values():
            if not certif_state.verified:
                self._users_cache[certif_state.content.user_id] = (now, certif_state.content)
        for certif_state in revoked_users_states.values():
            if not certif_state.verified:
                self._revoked_users_cache[certif_state.content.user_id] = (
                    now,
                    certif_state.content,
                )

        return (
            [state.content for state in users_states.values()],
            [state.content for state in revoked_users_states.values()],
            [state.content for state in devices_states.values()],
        )


_PyTrustchainContext = TrustchainContext
if not TYPE_CHECKING:
    try:
        from libparsec.types import TrustchainContext as _RsTrustchainContext
    except:
        pass
    else:
        TrustchainContext = _RsTrustchainContext

File: oxidation/libparsec/crates/core/src/trustchain.rs
// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS

use std::cell::RefCell;
use std::collections::HashMap;

use crate::{build_signature_path, TrustchainError, TrustchainResult};
use libparsec_crypto::VerifyKey;
use libparsec_protocol::authenticated_cmds::v2::user_get::Trustchain;
use libparsec_types::{
    CertificateSignerOwned, CertificateSignerRef, DateTime, DeviceCertificate, DeviceID,
    RevokedUserCertificate, TimeProvider, UserCertificate, UserID, UserProfile,
};

#[derive(Debug)]
struct CertifState<'a, T> {
    certif: &'a [u8],
    content: T,
    verified: bool,
}

pub struct TrustchainContext {
    root_verify_key: VerifyKey,
    time_provider: TimeProvider,
    cache_validity: i64,
    _users_cache: HashMap<UserID, (DateTime, UserCertificate)>,
    _devices_cache: HashMap<DeviceID, (DateTime, DeviceCertificate)>,
    _revoked_users_cache: HashMap<UserID, (DateTime, RevokedUserCertificate)>,
}

impl TrustchainContext {
    fn verify_created_by_device(
        &self,
        users_states: &HashMap<UserID, RefCell<CertifState<UserCertificate>>>,
        revoked_users_states: &HashMap<UserID, RefCell<CertifState<RevokedUserCertificate>>>,
        author_device: &DeviceCertificate,
        verified_data: (&UserID, DateTime),
        author: &DeviceID,
        sign_chain: &[String],
    ) -> TrustchainResult<()> {
        // Author is either admin or signing one of it own devices
        if author_device.device_id.user_id() != verified_data.0 {
            match users_states.get(author.user_id()) {
                Some(certif_state)
                    if certif_state.borrow().content.profile != UserProfile::Admin =>
                {
                    return Err(TrustchainError::InvalidSignatureGiven {
                        path: build_signature_path(sign_chain),
                        user_id: author.user_id().clone(),
                    })
                }
                None => {
                    return Err(TrustchainError::MissingUserCertificate {
                        path: build_signature_path(sign_chain),
                        user_id: author.user_id().clone(),
                    })
                }
                _ => (),
            }
        }

        // Also make sure author wasn't revoked at creation time
        if let Some(author_revoked_user) = revoked_users_states.get(author.user_id()) {
            let author_revoked_user = &author_revoked_user.borrow().content;

            if verified_data.1 > author_revoked_user.timestamp {
                return Err(TrustchainError::SignaturePosteriorUserRevocation {
                    path: build_signature_path(sign_chain),
                    verified_timestamp: verified_data.1,
                    user_timestamp: author_revoked_user.timestamp,
                });
            }
        }

        Ok(())
    }

    fn recursive_verify_device(
        &self,
        users_states: &HashMap<UserID, RefCell<CertifState<UserCertificate>>>,
        devices_states: &HashMap<DeviceID, RefCell<CertifState<DeviceCertificate>>>,
        revoked_users_states: &HashMap<UserID, RefCell<CertifState<RevokedUserCertificate>>>,
        device_id: &DeviceID,
        sign_chain: &mut Vec<String>,
    ) -> TrustchainResult<DeviceCertificate> {
        let device_id_str = device_id.to_string();

        if sign_chain.contains(&device_id_str) {
            sign_chain.push(device_id_str);
            return Err(TrustchainError::InvalidSignatureLoopDetected {
                path: build_signature_path(sign_chain),
            });
        }

        sign_chain.push(device_id_str);

        let state = devices_states
            .get(device_id)
            .ok_or(TrustchainError::MissingDeviceCertificate {
                path: build_signature_path(sign_chain),
                device_id: device_id.clone(),
            })?
            .borrow();

        match &state.content.author {
            CertificateSignerOwned::Root => DeviceCertificate::verify_and_load(
                state.certif,
                &self.root_verify_key,
                CertificateSignerRef::Root,
                None,
            )
            .map_err(|exc| TrustchainError::InvalidCertificate {
                path: build_signature_path(sign_chain),
                exc: exc.to_string(),
            }),
            CertificateSignerOwned::User(author) => {
                let author_device = self.recursive_verify_device(
                    users_states,
                    devices_states,
                    revoked_users_states,
                    author,
                    sign_chain,
                )?;

                let verified = DeviceCertificate::verify_and_load(
                    state.certif,
                    &author_device.verify_key,
                    CertificateSignerRef::User(&author_device.device_id),
                    None,
                )
                .map_err(|exc| TrustchainError::InvalidCertificate {
                    path: build_signature_path(sign_chain),
                    exc: exc.to_string(),
                })?;

                self.verify_created_by_device(
                    users_states,
                    revoked_users_states,
                    &author_device,
                    (verified.device_id.user_id(), verified.timestamp),
                    author,
                    sign_chain,
                )?;

                Ok(verified)
            }
        }
    }

    fn verify_user(
        &self,
        users_states: &HashMap<UserID, RefCell<CertifState<UserCertificate>>>,
        devices_states: &HashMap<DeviceID, RefCell<CertifState<DeviceCertificate>>>,
        revoked_users_states: &HashMap<UserID, RefCell<CertifState<RevokedUserCertificate>>>,
        unverified_content: &UserCertificate,
        certif: &[u8],
    ) -> TrustchainResult<UserCertificate> {
        let user_id = &unverified_content.user_id;

        match &unverified_content.author {
            CertificateSignerOwned::Root => UserCertificate::verify_and_load(
                certif,
                &self.root_verify_key,
                CertificateSignerRef::Root,
                None,
                None,
            )
            .map_err(|exc| TrustchainError::InvalidCertificate {
                path: build_signature_path(&[format!("{user_id}'s creation")]),
                exc: exc.to_string(),
            }),
            CertificateSignerOwned::User(author) if author.user_id() == user_id => {
                Err(TrustchainError::InvalidSelfSignedUserCertificate {
                    user_id: user_id.clone(),
                })
            }
            CertificateSignerOwned::User(author) => {
                let mut sign_chain = vec![format!("{user_id}'s creation")];

                let author_device = self.recursive_verify_device(
                    users_states,
                    devices_states,
                    revoked_users_states,
                    author,
                    &mut sign_chain,
                )?;

                let verified = UserCertificate::verify_and_load(
                    certif,
                    &author_device.verify_key,
                    CertificateSignerRef::User(&author_device.device_id),
                    None,
                    None,
                )
                .map_err(|exc| TrustchainError::InvalidCertificate {
                    path: build_signature_path(&sign_chain),
                    exc: exc.to_string(),
                })?;

                self.verify_created_by_device(
                    users_states,
                    revoked_users_states,
                    &author_device,
                    (&verified.user_id, verified.timestamp),
                    author,
                    &sign_chain,
                )?;

                Ok(verified)
            }
        }
    }

    fn verify_revoked_user(
        &self,
        users_states: &HashMap<UserID, RefCell<CertifState<UserCertificate>>>,
        devices_states: &HashMap<DeviceID, RefCell<CertifState<DeviceCertificate>>>,
        revoked_users_states: &HashMap<UserID, RefCell<CertifState<RevokedUserCertificate>>>,
        unverified_content: &RevokedUserCertificate,
        certif: &[u8],
    ) -> TrustchainResult<RevokedUserCertificate> {
        let author = &unverified_content.author;
        let user_id = &unverified_content.user_id;

        if author.user_id() == user_id {
            Err(
                TrustchainError::InvalidSelfSignedUserRevocationCertificate {
                    user_id: user_id.clone(),
                },
            )
        } else {
            let mut sign_chain = vec![format!("{user_id}'s revocation")];

            let author_device = self.recursive_verify_device(
                users_states,
                devices_states,
                revoked_users_states,
                author,
                &mut sign_chain,
            )?;

            let verified = RevokedUserCertificate::verify_and_load(
                certif,
                &author_device.verify_key,
                &author_device.device_id,
                None,
            )
            .map_err(|exc| TrustchainError::InvalidCertificate {
                path: build_signature_path(&sign_chain),
                exc: exc.to_string(),
            })?;

            self.verify_created_by_device(
                users_states,
                revoked_users_states,
                &author_device,
                (&verified.user_id, verified.timestamp),
                author,
                &sign_chain,
            )?;

            Ok(verified)
        }
    }
}

impl TrustchainContext {
    pub fn new(
        root_verify_key: VerifyKey,
        time_provider: TimeProvider,
        cache_validity: i64,
    ) -> Self {
        Self {
            root_verify_key,
            time_provider,
            cache_validity,
            _users_cache: HashMap::new(),
            _devices_cache: HashMap::new(),
            _revoked_users_cache: HashMap::new(),
        }
    }

    pub fn cache_validity(&self) -> i64 {
        self.cache_validity
    }

    pub fn invalidate_user_cache(&mut self, user_id: &UserID) {
        self._users_cache.remove(user_id);
    }

    pub fn get_user(&self, user_id: &UserID) -> Option<&UserCertificate> {
        let now = self.time_provider.now();

        match self._users_cache.get(user_id) {
            Some((cached_on, verified_user))
                if (now - *cached_on).num_seconds() < self.cache_validity =>
            {
                Some(verified_user)
            }
            _ => None,
        }
    }

    pub fn get_revoked_user(&self, user_id: &UserID) -> Option<&RevokedUserCertificate> {
        let now = self.time_provider.now();

        match self._revoked_users_cache.get(user_id) {
            Some((cached_on, verified_revoked_user))
                if (now - *cached_on).num_seconds() < self.cache_validity =>
            {
                Some(verified_revoked_user)
            }
            _ => None,
        }
    }

    pub fn get_device(&self, device_id: &DeviceID) -> Option<&DeviceCertificate> {
        let now = self.time_provider.now();

        match self._devices_cache.get(device_id) {
            Some((cached_on, verified_device))
                if (now - *cached_on).num_seconds() < self.cache_validity =>
            {
                Some(verified_device)
            }
            _ => None,
        }
    }

    pub fn load_trustchain(
        &mut self,
        users: &[Vec<u8>],
        revoked_users: &[Vec<u8>],
        devices: &[Vec<u8>],
    ) -> TrustchainResult<(
        Vec<UserCertificate>,
        Vec<RevokedUserCertificate>,
        Vec<DeviceCertificate>,
    )> {
        let mut users_states = HashMap::new();
        let mut devices_states = HashMap::new();
        let mut revoked_users_states = HashMap::new();

        // Deserialize the certificates and filter the ones we already have in cache
        for certif in devices {
            let unverified_device =
                DeviceCertificate::unsecure_load(certif).unwrap_or_else(|_| unreachable!());
            match self.get_device(&unverified_device.device_id) {
                Some(verified_device) => {
                    devices_states.insert(
                        verified_device.device_id.clone(),
                        RefCell::new(CertifState {
                            certif,
                            content: verified_device.clone(),
                            verified: true,
                        }),
                    );
                }
                _ => {
                    devices_states.insert(
                        unverified_device.device_id.clone(),
                        RefCell::new(CertifState {
                            certif,
                            content: unverified_device,
                            verified: false,
                        }),
                    );
                }
            }
        }

        for certif in users {
            let unverified_user =
                UserCertificate::unsecure_load(certif).unwrap_or_else(|_| unreachable!());
            match self.get_user(&unverified_user.user_id) {
                Some(verified_user) => {
                    users_states.insert(
                        verified_user.user_id.clone(),
                        RefCell::new(CertifState {
                            certif,
                            content: verified_user.clone(),
                            verified: true,
                        }),
                    );
                }
                _ => {
                    users_states.insert(
                        unverified_user.user_id.clone(),
                        RefCell::new(CertifState {
                            certif,
                            content: unverified_user,
                            verified: false,
                        }),
                    );
                }
            }
        }

        for certif in revoked_users {
            let unverified_revoked_user =
                RevokedUserCertificate::unsecure_load(certif).unwrap_or_else(|_| unreachable!());
            match self.get_revoked_user(&unverified_revoked_user.user_id) {
                Some(verified_revoked_user) => {
                    revoked_users_states.insert(
                        verified_revoked_user.user_id.clone(),
                        RefCell::new(CertifState {
                            certif,
                            content: verified_revoked_user.clone(),
                            verified: true,
                        }),
                    );
                }
                _ => {
                    revoked_users_states.insert(
                        unverified_revoked_user.user_id.clone(),
                        RefCell::new(CertifState {
                            certif,
                            content: unverified_revoked_user,
                            verified: false,
                        }),
                    );
                }
            }
        }

        // Verified what need to be and populate the cache with them
        // RefCell is intended to be used here because it is borrowed twice
        for certif_state in devices_states.values() {
            let verified = certif_state.borrow().verified;
            if !verified {
                let content = self.recursive_verify_device(
                    &users_states,
                    &devices_states,
                    &revoked_users_states,
                    &certif_state.borrow().content.device_id,
                    &mut vec![],
                )?;
                certif_state.borrow_mut().content = content;
            }
        }
        for certif_state in users_states.values() {
            let verified = certif_state.borrow().verified;
            if !verified {
                let content = self.verify_user(
                    &users_states,
                    &devices_states,
                    &revoked_users_states,
                    &certif_state.borrow().content,
                    certif_state.borrow().certif,
                )?;
                certif_state.borrow_mut().content = content;
            }
        }
        for certif_state in revoked_users_states.values() {
            let verified = certif_state.borrow().verified;
            if !verified {
                let content = self.verify_revoked_user(
                    &users_states,
                    &devices_states,
                    &revoked_users_states,
                    &certif_state.borrow().content,
                    certif_state.borrow().certif,
                )?;
                certif_state.borrow_mut().content = content;
            }
        }

        // Finally populate the cache
        let now = self.time_provider.now();
        devices_states
            .values()
            .map(|state| state.borrow())
            .filter(|state| !state.verified)
            .for_each(|state| {
                self._devices_cache.insert(
                    state.content.device_id.clone(),
                    (now, state.content.clone()),
                );
            });
        users_states
            .values()
            .map(|state| state.borrow())
            .filter(|state| !state.verified)
            .for_each(|state| {
                self._users_cache
                    .insert(state.content.user_id.clone(), (now, state.content.clone()));
            });
        revoked_users_states
            .values()
            .map(|state| state.borrow())
            .filter(|state| !state.verified)
            .for_each(|state| {
                self._revoked_users_cache
                    .insert(state.content.user_id.clone(), (now, state.content.clone()));
            });

        let verified_users = users_states
            .into_values()
            .map(|state| state.into_inner().content)
            .collect();
        let verified_revoked_users = revoked_users_states
            .into_values()
            .map(|state| state.into_inner().content)
            .collect();
        let verified_devices = devices_states
            .into_values()
            .map(|state| state.into_inner().content)
            .collect();

        Ok((verified_users, verified_revoked_users, verified_devices))
    }

    pub fn load_user_and_devices(
        &mut self,
        mut trustchain: Trustchain,
        user_certif: Vec<u8>,
        revoked_user_certif: Option<Vec<u8>>,
        devices_certifs: Vec<Vec<u8>>,
        expected_user_id: Option<&UserID>,
    ) -> TrustchainResult<(
        UserCertificate,
        Option<RevokedUserCertificate>,
        Vec<DeviceCertificate>,
    )> {
        let UserCertificate { user_id, .. } =
            UserCertificate::unsecure_load(&user_certif).unwrap_or_else(|_| unreachable!());

        if let Some(expected_user_id) = expected_user_id {
            if expected_user_id != &user_id {
                return Err(TrustchainError::UnexpectedCertificate {
                    expected: expected_user_id.clone(),
                    got: user_id,
                });
            }
        }

        trustchain.users.push(user_certif);
        trustchain.devices.extend(devices_certifs);

        let revoked_user_certif_is_some = match revoked_user_certif {
            Some(revoked_user_certif) => {
                trustchain.revoked_users.push(revoked_user_certif);
                true
            }
            None => false,
        };

        let (verified_users, verified_revoked_users, verified_devices) = self.load_trustchain(
            &trustchain.users,
            &trustchain.revoked_users,
            &trustchain.devices,
        )?;

        Ok((
            verified_users
                .into_iter()
                .find(|user| user.user_id == user_id)
                .unwrap_or_else(|| unreachable!()),
            if revoked_user_certif_is_some {
                verified_revoked_users
                    .into_iter()
                    .find(|user| user.user_id == user_id)
            } else {
                None
            },
            verified_devices
                .into_iter()
                .filter(|device| device.device_id.user_id() == &user_id)
                .collect(),
        ))
    }
}
